File: node/src/exchange-rate/exchange-rate.service.ts
import { Injectable } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { ExchangeRate } from './entities/exchange-rate';
import { InjectRepository } from '@nestjs/typeorm';
import { DataSource, Repository } from 'typeorm';
import { firstValueFrom } from 'rxjs/internal/firstValueFrom';
import { IExchangeRatesDto } from './dto/exchange-rates.dto';
import { AxiosResponse } from 'axios';

const API_URL: string =
  'https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies/usd.json';
const API_URL_FALLBACK: string =
  'https://latest.currency-api.pages.dev/v1/currencies/usd.json';

@Injectable()
export class ExchangeRateService {
  constructor(
    @InjectRepository(ExchangeRate)
    private exchangeRateRepository: Repository<ExchangeRate>,
    private dataSource: DataSource,
  ) { }

  findAll(): Promise<ExchangeRate[]> {
    return this.exchangeRateRepository.find();
  }

  async getTodayFiveRates(): Promise<Promise<ExchangeRate[]> | null> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    //first we try the db
    let rates = await this._getTodayFiveRatesDb(today);
    if (rates && rates.length) {
      return rates;
    }

    //now we try the api
    rates = await this._getTodayFiveRatesApi(today, API_URL);
    if (rates && rates.length) {
      return rates;
    }

    //maybe the fallback api
    rates = await this._getTodayFiveRatesApi(today, API_URL_FALLBACK);
    if (rates && rates.length) {
      return rates;
    }

    return null;
  }

  private async _getTodayFiveRatesDb(
    today: Date,
  ): Promise<ExchangeRate[] | null> {
    return this.exchangeRateRepository.find({ where: { CreatedAt: today } });
  }

  private async _getTodayFiveRatesApi(
    today: Date,
    url: string,
  ): Promise<Promise<ExchangeRate[]> | null> {
    try {
      const httpService = new HttpService();
      const response: AxiosResponse = await firstValueFrom(
        httpService.get(url),
      );
      const ratesDto: IExchangeRatesDto = response.data as IExchangeRatesDto;
      const rates: ExchangeRate[] = this._ratesFromDto(ratesDto, today);
      await this._saveToDb(rates);
      return rates;
    } catch (error) {
      console.log(error);
      return null;
    }
  }

  private async _saveToDb(rows: ExchangeRate[]): Promise<boolean> {
    try {
      //batch insert for optimization, we may be in a huge importing operation...
      await this.dataSource
        .createQueryBuilder()
        .insert()
        .into(ExchangeRate)
        .values(rows)
        .execute();
    } catch (error) {
      console.log(error);
      return false;
    }
    return true;
  }

  private _ratesFromDto(
    ratesDto: IExchangeRatesDto,
    today: Date,
  ): ExchangeRate[] {
    const date = new Date(ratesDto.date);
    const rates: ExchangeRate[] = [
      {
        Id: 0,
        Date: date,
        CreatedAt: today,
        Abbreviation: 'BRL',
        Rate: ratesDto.usd.brl,
      },
      {
        Id: 0,
        Date: date,
        CreatedAt: today,
        Abbreviation: 'CAD',
        Rate: ratesDto.usd.cad,
      },
      {
        Id: 0,
        Date: date,
        CreatedAt: today,
        Abbreviation: 'EUR',
        Rate: ratesDto.usd.eur,
      },
      {
        Id: 0,
        Date: date,
        CreatedAt: today,
        Abbreviation: 'GBP',
        Rate: ratesDto.usd.gbp,
      },
      {
        Id: 0,
        Date: date,
        CreatedAt: today,
        Abbreviation: 'MXN',
        Rate: ratesDto.usd.mxn,
      },
    ];

    return rates;
  }
}

File: backend/_others_/cs_dapper/Services/ExchangeRateService.cs
using cssharp.Models;
using cssharp.Repositories;
using System.Text.Json;

namespace cssharp.Services
{
  public interface IExchangeRateService
  {
    public Task<IEnumerable<ExchangeRate>> FindAll();

    public Task<IEnumerable<ExchangeRate>> GetToday5Rates();
  }

  public class ExchangeRateService : IExchangeRateService
  {
    private const string URL_API = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies/usd.json";
    private const string URL_API_2 = "https://latest.currency-api.pages.dev/v1/currencies/usd.json";

    IExchangeRateRepository _repository;
    HttpClient _httpClient;

    public ExchangeRateService(IExchangeRateRepository repository, HttpClient httpClient)
    {
      _repository = repository;
      _httpClient = httpClient;
    }

    public async Task<IEnumerable<ExchangeRate>> FindAll()
    {
      return await _repository.FindAll();
    }

    public async Task<IEnumerable<ExchangeRate>> GetToday5Rates()
    {
      var rates = await _GetToday5RatesDb();
      if (rates != null && rates.Any())
        return rates;
      rates = await _GetToday5RatesApi(URL_API);
      if (rates != null && rates.Any())
      {
        await _SaveRatesAsync(rates);
        return rates;
      }
      rates = await _GetToday5RatesApi(URL_API_2);
      await _SaveRatesAsync(rates);
      return rates;
    }

    private async Task _SaveRatesAsync(IEnumerable<ExchangeRate> rates)
    {
      foreach (var rate in rates)
      {
        await _repository.Save(rate);
      }
    }

    private async Task<IEnumerable<ExchangeRate>> _GetToday5RatesApi(string url)
    {
      HttpResponseMessage response = await _httpClient.GetAsync(url);
      if (!response.IsSuccessStatusCode)
        return Enumerable.Empty<ExchangeRate>();
      string body = await response.Content.ReadAsStringAsync();

      using JsonDocument doc = JsonDocument.Parse(body);
      JsonElement root = doc.RootElement;
      DateTime date = root.GetProperty("date").GetDateTime();
      JsonElement usd = root.GetProperty("usd");

      ExchangeRate[] rates = new ExchangeRate[5];
      rates[0] = _GetRateFromJson(usd, "brl", date!);
      rates[1] = _GetRateFromJson(usd, "eur", date!);
      rates[2] = _GetRateFromJson(usd, "cad", date!);
      rates[3] = _GetRateFromJson(usd, "mxn", date!);
      rates[4] = _GetRateFromJson(usd, "gbp", date!);

      return rates;
    }

    private ExchangeRate _GetRateFromJson(JsonElement usd, string abbreviation, DateTime date)
    {
      var rate = new ExchangeRate()
      {
        Id = 0,
        CreatedAt = DateTime.Now,
        Date = date,
        Abbreviation = abbreviation.ToUpper(),
        Rate = usd.GetProperty(abbreviation).GetDecimal(),
      };
      return rate;
    }

    private async Task<IEnumerable<ExchangeRate>> _GetToday5RatesDb()
    {
      return await _repository.FindByDate(DateTime.Now);
    }
  }
}
